Products/Zuul/catalog/model_catalog_tool.py

```diff
@@ -0,0 +1,317 @@
+
+log = logging.getLogger("model_catalog_tool")
+
+
+class SearchResults(object):
+
+    def __init__(self, results, total, hash_, areBrains=True):
+        self.results = results
+        self.total = total
+        self.hash_ = hash_
+        self.areBrains = areBrains
+
+    def __hash__(self):
+        return self.hash_
+
+    def __iter__(self):
+        return self.results
+
+
+class ModelCatalogTool(object):
+    """ Search the model catalog """
+
+    implements(IModelCatalogTool)
+
+    def __init__(self, context):
+        self.context = context
+        self.model_catalog = getUtility(IModelCatalog)
+
+    def is_model_catalog_enabled(self):
+        return self.model_catalog.searcher is not None
+
+    def _build_query(self, types=(), paths=(), depth=None, query=None, filterPermissions=True, globFilters=None):
+        """
+        Build and AdvancedQuery query
+
+        @params types: list/tuple of values for objectImplements field
+        @params globFilters: dict with user passed field: value filters
+        @params query: AdvancedQuery passed by the user. Most of the time None
+        @param filterPermissions: Boolean indicating whether to check for user perms or not
+
+        @return: tuple (AdvancedQuery query, not indexed filters dict)
+        """
+        available_indexes = self.model_catalog.searcher.get_indexes()
+        not_indexed_user_filters = {} # Filters that use not indexed fields
+
+        user_filters_query = None
+        types_query = None
+        paths_query = None
+        permissions_query = None
+
+        partial_queries = []
+
+        if query:
+            partial_queries.append(query)
+
+        # Build query from filters passed by user
+        if globFilters:
+            for key, value in globFilters.iteritems():
+                if key in available_indexes:
+                    if user_filters_query:
+                        user_filters_query = And(query, MatchRegexp(key, '*%s*' % value))
+                    else:
+                        user_filters_query = MatchRegexp(key, '*%s*' % value)
+                else:
+                    not_indexed_user_filters[key] = value
+
+        if user_filters_query:
+            partial_queries.append(user_filters_query)
+
+        # Build the objectImplements query
+        if not isinstance(types, (tuple, list)):
+            types = (types,)
+        types_query_list = [ Eq('objectImplements', dottedname(t)) for t in types ]
+        if len(types_query_list) > 1:
+            types_query = Or(*types_query_list)
+        else:
+            types_query = types_query_list[0]
+
+        if types_query:
+            partial_queries.append(types_query)
+
+        # Build query for paths
+        if not paths:
+            paths = ('/'.join(self.context.getPhysicalPath()) + '*', )
+        elif isinstance(paths, basestring):
+            paths = (paths,)
+
+        """  OLD CODE. Why this instead of In?  What do we need depth for?
```

Comment:
I'm not sure we currently use depth, although I can't remember. It will search at a specific depth in the path, e.g. "Get anything 3 levels below /zport/dmd/Devices named Server".